indexer/ai/src/main/java/com/google/cloudsearch/ai/BaseAISkill.java

```diff
@@ -0,0 +1,139 @@
+package com.google.cloudsearch.ai;
+
+import com.google.cloudsearch.exceptions.InvalidConfigException;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import java.util.*;
+
+/**
+ * Abstract class BaseAISkill implements the common functionality required for the different skills.
+ */
+public abstract class BaseAISkill implements AISkill {
+
+    private String aiSkillName = "";
+    private List<OutputMapping> outputMappings = new ArrayList<OutputMapping>();
+    private Map<String, Boolean> schemaInfo = new HashMap<String, Boolean>();
+    private Boolean isSchemaInfoStored = false;
+
+    /**
+     * Parses the CloudsSearch Schema and stores the object and property names
+     * @param schema    CloudSearch schema
+     */
+    private void storeSchemaInfo(JSONObject schema){
+        JSONArray schemaObjects = (JSONArray) schema.get(Constants.CONFIG_SCHEMA_OBJECT_DEFINITIONS);
+        for(Object schemaObj : schemaObjects){
+            JSONObject obj = (JSONObject) schemaObj;
+            String objName = (String) obj.get(Constants.CONFIG_SCHEMA_NAME);
+            JSONArray propertyDefinitions = (JSONArray) obj.get(Constants.CONFIG_SCHEMA_PROPERTY_DEFINITIONS);
+            for(Object propertyObj : propertyDefinitions){
+                JSONObject property = (JSONObject) propertyObj;
+                String propertyName = (String) property.get(Constants.CONFIG_SCHEMA_NAME);
+                schemaInfo.put(objName+"."+propertyName, true);
+            }
+        }
+        this.isSchemaInfoStored = true;
+    }
+
+    /**
+     * Helper function to validate property names in config
+     * @param propertyName  object name + property name specified in configuration
+     * @param schema    CloudSearch schema
+     * @return
+     */
+    private boolean isValidPropertyName(String propertyName, JSONObject schema){
```

Comment:
spacing between ) and { is off at multiple places.

Reply:
sure, I will fix this with the changes for the next PR

Reply:
you can amend the last commit as well

Reply:
Oh, I actually made lot of changes now, as per Ashish's comments in the latest PR, so is it fine if I push all that together with PR #11?

Reply:
just keep changes relevant to this PR only, you can amend the commit